src/golang/lib/workflow/artifact/artifact.go

```diff
@@ -45,49 +49,69 @@ type Artifact interface {
 }
 
 type ArtifactImpl struct {
-	id           uuid.UUID
+	// This is the ID that will be stored in our database. It is the canonical handle
+	// to this artifact throughout our system.
+	id uuid.UUID
+
+	// This is a special ID that also encodes important structural/parameter information.
+	// This can be used as a unique handle to a specific artifact **for a specific dag execution**.
+	logicalID uuid.UUID
+
 	name         string
 	description  string
 	artifactType artifact.Type
 
-	contentPath  string
-	metadataPath string
+	execPaths *utils.ExecPaths
 
-	resultWriter     artifact_result.Writer
-	resultID         uuid.UUID
-	resultsPersisted bool
+	resultWriter artifact_result.Writer
+	resultID     uuid.UUID
+
+	// If this is not nil, this artifact should be written to the cache.
+	// An artifact cannot be both cache-aware and persisted.
+	previewArtifactCacheManager PreviewCacheManager
+	resultsPersisted            bool
 
 	storageConfig *shared.StorageConfig
 	db            database.Database
 }
 
 func NewArtifact(
+	logicalID uuid.UUID,
```

Comment:
Let's rename logical id to something like `cacheSignature`. Logical id now contains parameter value that's determined during each run, so it doesn't really belong to a logical property.

Reply:
hm what about `previewCacheKey` or `artifactSignature`? I feel like "key" is more accurate term here, but I also don't mind the "signature" terminology.